cpp/src/io/parquet/writer_impl.cu

```diff
@@ -531,7 +536,43 @@ std::vector<schema_tree_node> construct_schema_tree(
         }
       };
 
-      if (col->type().id() == type_id::STRUCT) {
+      auto last_list_child = [](cudf::detail::LinkedColPtr col) {
+        if (col->type().id() != type_id::LIST) { return false; }
+        auto const child_col_type =
+          col->children[lists_column_view::child_column_index]->type().id();
+        return child_col_type == type_id::INT8 or child_col_type == type_id::UINT8;
+      };
+
+      // There is a special case for a list<int8> column with one byte column child. This column can
+      // have a special flag that indicates we write this out as binary instead of a list. This is a
+      // more efficient storage mechanism for a single-depth list of bytes, but is a departure from
+      // original cuIO behavior so it is locked behind the option. If the option is selected on a
+      // column that isn't a single-depth list<int8> the code will throw.
+      if (col_meta.is_enabled_output_as_binary() && last_list_child(col)) {
+        CUDF_EXPECTS(col_meta.num_children() == 2 or col_meta.num_children() == 0,
+                     "Binary column's corresponding metadata should have zero or two children!");
+        if (col_meta.num_children() > 0) {
+          auto const data_col_type =
+            col->children[lists_column_view::child_column_index]->type().id();
+          CUDF_EXPECTS(data_col_type == type_id::INT8 || data_col_type == type_id::UINT8,
+                       "Binary column's type must be INT8 or UINT8!");
+
+          CUDF_EXPECTS(col->children[lists_column_view::child_column_index]->children.size() == 0,
+                       "Binary column must not be nested!");
+        }
+
+        schema_tree_node col_schema{};
+        col_schema.type            = Type::BYTE_ARRAY;
+        col_schema.converted_type  = ConvertedType::UNKNOWN;
+        col_schema.stats_dtype     = statistics_dtype::dtype_byte_array;
+        col_schema.repetition_type = col_nullable ? OPTIONAL : REQUIRED;
+        col_schema.name = (schema[parent_idx].name == "list") ? "element" : col_meta.get_name();
+        col_schema.parent_idx  = parent_idx;
+        col_schema.leaf_column = col;
```

Comment:
Nitpick: Looks like `schema_tree_node` needs a constructor. :]

I would've recommended aggregate initialization, but I'm not sure how appropriate it would be, since it's initializing parent members:
```suggestion
        schema_tree_node col_schema {
          .type            = Type::BYTE_ARRAY,
          .converted_type  = ConvertedType::UNKNOWN,
          .stats_dtype     = statistics_dtype::dtype_byte_array,
          .repetition_type = col_nullable ? OPTIONAL : REQUIRED,
          .name = (schema[parent_idx].name == "list") ? "element" : col_meta.get_name(),
          .parent_idx  = parent_idx,
          .leaf_column = col
        };
```

My personal preference would be to avoid construct-then-initialize. But this is purely stylistic. Please feel free to ignore.

Reply:
Added an issue for you :)

https://github.com/rapidsai/cudf/issues/11395